pkg/hub/submarinerconfig/controller.go

```diff
@@ -0,0 +1,269 @@
+		// ignore bad format key
+		return nil
+	}
+
+	// reconcile the managed cluster to handle creating submariner config before creating managed cluster
+	if namespace == "" {
+		configs, err := c.configClient.SubmarineraddonV1alpha1().SubmarinerConfigs(name).List(ctx, metav1.ListOptions{})
+		if err != nil {
+			return err
+		}
+		switch len(configs.Items) {
+		case 0:
+			return nil
+		case 1:
+			return c.syncConfig(ctx, &configs.Items[0])
+		default:
+			return fmt.Errorf("one more than submariner configs are found for %q", name)
+		}
+	}
+
+	config, err := c.configClient.SubmarineraddonV1alpha1().SubmarinerConfigs(namespace).Get(ctx, name, metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		// config is not found, could have been deleted, do nothing.
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	return c.syncConfig(ctx, config)
+}
+
+func (c *submarinerConfigController) syncConfig(ctx context.Context, config *configv1alpha1.SubmarinerConfig) error {
+	if config.DeletionTimestamp.IsZero() {
+		hasFinalizer := false
+		for i := range config.Finalizers {
+			if config.Finalizers[i] == submarinerConfigFinalizer {
+				hasFinalizer = true
+				break
+			}
+		}
+		if !hasFinalizer {
+			config.Finalizers = append(config.Finalizers, submarinerConfigFinalizer)
+			_, err := c.configClient.SubmarineraddonV1alpha1().SubmarinerConfigs(config.Namespace).Update(ctx, config, metav1.UpdateOptions{})
+			return err
+		}
+	}
+
+	// config is deleting, we remove its related resources
+	if !config.DeletionTimestamp.IsZero() {
+		if err := c.cleanUpSubmarinerClusterEnv(ctx, config); err != nil {
+			return err
+		}
+		return c.removeConfigFinalizer(ctx, config)
+	}
+
+	if config.Spec.CredentialsSecret == nil {
+		// no platform credentials, the submariner cluster environment neet not to be prepared
+		return nil
+	}
+
+	managedCluster, err := c.clusterClient.ClusterV1().ManagedClusters().Get(context.TODO(), config.Namespace, metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		// related managed cluster is not found, we cannot prepare the submariner cluster environment for it
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	// prepare submariner cluster environment
+	errs := []error{}
+	cloudProvider, preparedErr := cloud.GetCloudProvider(c.kubeClient, c.workClient, c.eventRecorder, managedCluster, config)
+	if preparedErr == nil {
+		preparedErr = cloudProvider.PrepareSubmarinerClusterEnv()
+	}
+
+	condition := metav1.Condition{
+		Type:    configv1alpha1.SubmarinerConfigConditionEnvPrepared,
+		Status:  metav1.ConditionTrue,
+		Reason:  "SubmarinerClusterEnvPrepared",
+		Message: "Submariner cluster environment was prepared",
+	}
+	switch {
+	case preparedErr == nil:
+		// add a finalizer to managed cluster to avoid the managed cluster is deleted in advance,
+		// because when we clean up the submariner cluster environment, we need the cluster informations
+		if err := c.addManagedClusterFinalizer(ctx, managedCluster); err != nil {
```

Comment:
we should think whether maintaining these info in status of config if better than adding finalizer in managedcluster. We may not want each addon to add a  finalizer.

Reply:
use status instead of finalizer